stingray/covariancespectrum.py

```diff
@@ -0,0 +1,230 @@
+                mid_bin = (band[0] + band[1]) / 2
+                energy_events_[mid_bin] = []
+
+                # Modify self.energy_events to form a band with one key
+                for key in list(self.energy_events.keys()):
+                    if key >= band[0] and key <= band[1]:
+                        energy_events_[mid_bin] += self.energy_events[key]
+                        del self.energy_events[key]
+
+            self.energy_events.update(energy_events_)
+
+    def _init_energy_covar(self):
+        """
+        Initialize the energy_covar dictionary for further computations.
+        """
+        # The dictionary with covariance spectrum for each energy bin
+        self.energy_covar = {}
+
+        # Initialize it with empty mapping
+        if self.band_interest is None:
+            for key in self.energy_events.keys():
+                self.energy_covar[key] = []
+        else:
+            for band in list(self.band_interest):
+                mid_bin = (band[0] + band[1]) / 2
+                self.energy_covar[mid_bin] = []
+
+    def _construct_energy_covar(self):
+        """Form the actual output covaraince dictionary and array."""
+        self._init_energy_covar()
+
+        xs_var = dict()
+
+        for energy in self.energy_covar.keys():
+            lc = Lightcurve.make_lightcurve(
+                    self.energy_events[energy], self.dt, tstart=self.min_time,
+                    tseg=self.max_time - self.min_time)
+
+            # Calculating timestamps for lc_ref
+            toa_ref = []
+            for key, value in self.energy_events.items():
+                if key >= self.ref_band_interest[0] and \
+                        key <= self.ref_band_interest[1]:
+                    if key != energy:
+                        toa_ref.extend(value)
+
+            toa_ref = np.array(sorted(toa_ref))
+
+            lc_ref = Lightcurve.make_lightcurve(
+                    toa_ref, self.dt, tstart=self.min_time,
+                    tseg=self.max_time - self.min_time)
+
+            assert len(lc.time) == len(lc_ref.time)
+
+            covar = self._compute_covariance(lc, lc_ref)
+
+            self.energy_covar[energy] = covar
+
+            # Calculate error
+            if self.error is None:
+                std = np.mean(lc_ref)**0.5
+            elif type(self.error) in [list, np.array]: # Array of numbers
+                std = np.mean(self.error)
+            else:  # Single float number
+                std = self.error
+
+            xs_var[energy] = np.var(lc_ref) - std**2  # Excess variance in ref band
+
+        self.unnorm_covar = np.vstack(self.energy_covar.items())
+
+        for key, value in self.energy_covar.items():
+            if not xs_var[key] > 0 :
+                utils.simon('xs_var must not take non-positive values.')
+            self.energy_covar[key] = value / (xs_var[key])**0.5
+
+        self.covar = np.vstack(self.energy_covar.items())
+
```

Comment:
One last thing missing: the error on the covariance (it's not numbered, but it's effectively Equation 4 in Wilkinson + Uttley 2009).